### PR TITLE
notify_socket.go: use sd_notify_barrier mechanism

### PR DESCRIPTION
**Pull request by indyjo**
Add synchronization to how runc reports container readiness to systemd.

It was previously possible for runc to terminate  immediately after sending the READY signal. This was sometimes too quick for systemd to find out the sender's cgroup. As a consequence, systemd would ignore the READY signal.

Fixes: #3293

### CODE REVIEW

File: notify_socket.go
Comment (kolyshkin) [inline notify_socket.go:217]: Technically it's not needed; you can disable the linter warning using something like `//nolint:errorlint // comparison with io.EOF is legit.`

File: notify_socket.go
Comment (kolyshkin) [inline notify_socket.go:195]: It's better to have complete sentences (starting with an uppercase letter, ending with a period) as comments. IOW I like the earlier form better.

File: notify_socket.go
Comment (indyjo) [reply notify_socket.go]: Done.

File: notify_socket.go
Comment (indyjo) [reply notify_socket.go]: I agree. Changed where there were no previous comments.

File: notify_socket.go
Comment (AkihiroSuda) [inline notify_socket.go:217]: Can we check the content of the byte too?

Comment (Alaric84) [reply notify_socket.go]: I'd stay as close to the systemd implementation [systemd/sd_notify_barrier](https://github.com/systemd/systemd/blob/5efbd0bf897a990ebe43d7dc69141d87c404ac9a/src/libsystemd/sd-daemon/sd-daemon.c#L555) with the given technical possibilities of go.
The content of the byte is not part of the "protocol" - so we should not care about it.

**Comment by kolyshkin** [comment]
Overall this looks good to me. Is there any way to have a test for this @indyjo ?

**Comment by indyjo** [comment]
What I can imagine is to write a test against a mocked NOTIFY_SOCKET server, i.e. a mocked systemd. Concerning how to do it I think I need some guidance. I could try adding a `notify_socket_test.go` alongside  `notify_socket.go`. I also noticed there are "bats tests" in `tests/integration`, but I'm completely unfamiliar with that framework. What do you think would be a good approach?

Also, my code is currently assuming that whoever listens on the other end of NOTIFY_SOCKET actually understands and honors the `BARRIER=1` protocol. This is certainly the case with recent systemd versions (the feature seems to be about two years old). Other implementations might not. For example, [sdnotify-proxy](https://github.com/coreos/sdnotify-proxy) is too simple. Currently, this would lead to `runc` running into the 5 second timeout and failing to start a container.

Is this behavior a) accepted or should I b) ignore the timeout case or should there be c) a command-line option to suppress the sd_notify_barrier feature?

**Comment by indyjo** [comment]
I have added a test (notify_socket_test.go). The question of how to deal with sd_notify_barrier timeouts is still open. `crun` waits for 30 seconds and then ignores the error ([see PR](https://github.com/containers/crun/pull/791)). `systemd-notify` waits for 5 seconds and then fails ([source](https://github.com/systemd/systemd/blob/e6f48be8d4432924028ca0d5ee5289626a2fd7bb/src/notify/notify.c#L277)).

File: notify_socket.go
Comment (kolyshkin) [inline notify_socket.go:187]: Docstring is supposed to start with the identifier, i.e. in this case it should be "sdNotifyBarrier performs synchronization with systemd ....".

File: notify_socket.go
Comment (kolyshkin) [inline notify_socket.go:183]: Docstring is supposed to start with the identifier, i.e. in this case it should be "errUnexpectedRead is ...."

File: notify_socket.go
Comment (kolyshkin) [inline notify_socket.go:203]: maybe use `unix.UnixRights` and `unix.Sendmsg` instead.

File: notify_socket.go
Comment (kolyshkin) [inline notify_socket.go:205]: This `err` is bare, e.g. something like EPERM, with no context.

Makes sense to wrap it into, say, `os.SyscallError`:
```go
return &os.SyscallError{Syscall: "sendmsg", Err: err}
```

File: notify_socket_test.go
Comment (kolyshkin) [inline notify_socket_test.go:105]: same here -- use `unix.` instead.

File: notify_socket.go
Comment (indyjo) [reply notify_socket.go]: Done.

File: notify_socket.go
Comment (indyjo) [reply notify_socket.go]: Done.

File: notify_socket.go
Comment (indyjo) [reply notify_socket.go]: Done.

File: notify_socket.go
Comment (indyjo) [reply notify_socket.go]: Done.

File: notify_socket_test.go
Comment (indyjo) [reply notify_socket_test.go]: Done.

File: notify_socket.go
Comment (kolyshkin) [inline notify_socket.go:167]: Hmm, wonder why we avoid using bytes.Buffer at all and don't just do something like
```go
_, err = client.WriteString(string(ready)+"\n")
```

Comment (kolyshkin) [reply notify_socket.go:167]: > Hmm, wonder why we avoid using bytes.Buffer at all 

Looking at git history, this was done in #1308 but I don't see why.

Comment (kolyshkin) [reply notify_socket.go:167]: > _, err = client.Write(append(ready, '\n'))

This one looks fine to me (most probably there's `\n` already anyway at that position), as we are not reusing the buffer anywhere.

Can you implement it (as a separate first commit, I guess)?

**Comment by kolyshkin** [comment]
> * reverse the order of commits (first remove bytes.Buffer usage, then add your changes)

I'm asking for this because it makes the git log -p smaller (and, consequently, it will be easier to figure out what was changed, especially a few years from now).

**Comment by indyjo** [comment]
Thanks for clarifying, @kolyshkin! The two commits are now in the preferred order. 

Should we discuss how to handle the case where the supervising process doesn't (correctly) handle the `sd_notify_barrier` mechanism and `runc` fails with a timeout? I'm leaning towards simply doing what `crun` does, although the 30s timeout seems extreme.

**Comment by kolyshkin** [comment]
> I'm leaning towards simply doing what `crun` does

I tend to agree.

> although the 30s timeout seems extreme.

I guess on an overloaded system 5s might not be enough, so the timeout should be way above the reasonable expected reply time.

But let's ask @giuseppe -- why you ended up with 30s timeout?

**Comment by giuseppe** [comment]
> But let's ask @giuseppe -- why you ended up with 30s timeout?

I've copied the timeout you set here: https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/systemd/common.go#L345  :smile: 

Although I've decided to ignore any error, so it is equivalent to the version without `sd_notify_barrier` where it would not raise an error and because I am not sure as well what a correct value would be for the timeout

**Comment by indyjo** [comment]
Committed the proposed solution (as a separate commit for your convenience, @kolyshkin). Wait 30s and warn but ignore on timeout. Just tell me and I'll merge the two commits into one.

**Comment by kolyshkin** [comment]
LGTM @indyjo; please squash the last two commits and let's get it merged (after 1.1.0 I guess)

**Comment by indyjo** [comment]
Here you go, @kolyshkin, and happy new year!

**Review by kolyshkin — DISMISSED** [review]
LGTM

Alas I am afraid it's too late for runc 1.1.

**Comment by kolyshkin** [comment]
@AkihiroSuda PTAL

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers @AkihiroSuda PTAL

**Comment by MofX** [comment]
Does it really make sense to implement the barrier-machanism inside of runc?
1. This will fail if systemd is too old (I guess this does not matter, because it only triggers a warning?)
2. For non-detached containers, if systemd-notify or any other binary, that uses sd_notify_barrier is used inside of the container it will fail, because the pipe and the barrier message is not relayed to systemd leading to a timeout in sd_notify_barrier.
3. For detached containers, sd_notify_barrier can or cannot timeout, depending on a race. If I don't miss anything, runc terminates, after it received the ready message (and relayed it to systemd). After the patch it will also do the barrier synchronization. A process using sd_notify_barrier will most likely work with this PR, if it is able to send the barrier message with it's own pipe fd before runc terminates. I think the recv from the socket with read (in contrast to recv) just closes open handles. But if runc terminates before it read the BUFFER message and the pipe fd, sd_notify_barrier will timeout.

**Comment by indyjo** [comment]
Hi @MofX, let me give my point of view. And sorry about not answering fo so long.

First of all, whether this patch ends up in an official release or not is not absolutely critical for me. The company I work for, KUKA, maintains their own patched version of `runc`. We have made good experiences so far, but of course we only test a very limited scenario.

> 1. This will fail if systemd is too old (I guess this does not matter, because it only triggers a warning?)

Probably true. The protocol basically mandates the use of the barrier, though. There is no way for a client to query whether it's supported.

Concerning your point 2 and 3:

I don't know enough about `runc` to understand the implications of detached and detached containers. It is true, however, that `systemd-notify` inside a container can still fail. That's because this patch doesn't address the barrier mechanism between runc and the container, but between runc and its supervising process.

Actually, we never observed any _timeout_ happening in `systemd-notify`. The way `systemd-notify` may fail is that sometimes `runc` has already closed the connection directly after receiving the READY state, causing a "Connection refused" when `systemd-notify` tries to establish the barrier. We usually disable the barrier mechanism for this reason, when running `systemd-notify` in a container, using the `--no-block` option. That's a separate issue, though.

**Comment by indyjo** [comment]
@kolyshkin May I ask if there's a chance of including this in v1.1.2?

**Comment by kolyshkin** [comment]
> @kolyshkin May I ask if there's a chance of including this in v1.1.2?

The target milestone is set for 1.2.0 (which I hope will be released some time in the summer). We can definitely do a backport to 1.1.x once this is merged, and so it will also be included into latest 1.1.x (whatever that would be).

We need more reviews on this, @opencontainers/runc-maintainers PTAL 🙏🏻

**Comment by kolyshkin** [comment]
@indyjo can you please rebase it (this will also re-run the CI on golang 1.18)?

**Comment by indyjo** [comment]
> @indyjo can you please rebase it (this will also re-run the CI on golang 1.18)?

@kolyshkin Technically, it's done. There's a failing check though: `ci / centos-7`.

**Comment by indyjo** [comment]
Friendly reminder that there's still a chance for this baby to be merged before its first birthday! 🎂
Is there anything I can help with? @kolyshkin @thaJeztah @cyphar @AkihiroSuda

**Review by AkihiroSuda — DISMISSED** [review]

**Comment by AkihiroSuda** [comment]
The last rebase was 6 months ago, could you rebase again to ensure CI is still green? Thanks

**Comment by indyjo** [comment]
Rebased and green. @AkihiroSuda @kolyshkin

**Review by AkihiroSuda — APPROVED** [review]

**Comment by indyjo** [comment]
@kolyshkin Just a friendly reminder

**Review by kolyshkin — APPROVED** [review]
LGTM

**Pull request merged**
